### ISSUE TITLE
Taxonomy: Scope and Task Breakdown

### ISSUE THREAD

**Issue by alexkim205** [PostHog/posthog#7589]
**Goal**: Users should understand events, actions & properties others have set-up (more details in the roadmap).

The below comes out of a meeting we (@rcmarron and I) had about what exactly we should be tackling for taxonomy. Tagging my team for context and feedback (@rcmarron @paolodamico @clarkus @pauldambra @Twixes @mariusandra @marcushyett-ph ).

## Scope

🦄  = frontend
🐴 = backend
🎨 = needs design guidance/mockups
🧠  = still needs clarification
🔻 = deprioritized

### 1. Helping people organize events and properties

**Automated Taxonomy**
* Metrics
	* Time metrics (last seen, created)
	* Event Volume (# of events across time range)
	* Query Volume (# of insight queries using this event/property)
	* Format (dates, auto-detected fields) 🧠 
* Only showing properties that are relevant/can be filtered on
	* Mapping events to properties
* How often is an event used in queries?
* Incomplete/bad event data validation
	* Every property has a type that could be required/optional
	* Property typing (string, boolean, null, enum, number)

**Manual Taxonomy**
* Verified + stale events
* Hiding events 🧠 (project level setting?)
* Event version history
* Set event property types
* Easy search and filtering
* Event (property) descriptions 🔻  _(I'd like to challenge that users wouldn't use this too much, since manually annotating all events/event properties can become a tedious task with no version control). If you need a property description, you should probably rename that property and/or event name/description._

**Misc.**
* User group + cohorts properties

### 2. How do we expose this to users?

* **Remove** `Properties Stats` (see Events Definitions Tab below for more information)
* **Rename** `Event Stats` to `Events Definitions`

`Events Definitions` Tab 🦄

<table>
<tr>
	<td>Status
	<td>Task
	<td>Complexity
	<td>Uncertainty
	<td>Expected
	<td>Worst-case
	<td>Notes
<tr>
	<td>
	<td>Link event in `Events Stats` to corresponding insight
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td>
<tr>
	<td>
	<td>Add sortable columns to table for taxonomic metrics (time, event volume, query volume)		
	<td>small
	<td>moderate
	<td>1d
	<td>1.5d
	<td>blocked on backend logic to calculate event stats
<tr>
	<td>
	<td>🎨 Use new taxonomic filter (allow filtering on events and event properties)
	<td>medium
	<td>low
	<td>2d
	<td>2.1d
	<td>How does this change how matching event properties (nested under event) are shown?
<tr>
	<td>
	<td>Active/Inactive (Verified/Unverified/Stale?) toggle per event row
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td>blocked on new event update api change
<tr>
	<td>
	<td>Add section for configuring handling of badly formed data.
	<td>medium
	<td>moderate
	<td>2d
	<td>3d
	<td>"ignore" or "capture and tag"
</table>

Event Properties 🦄  (move `Properties Stats` under corresponding event in `Events Definitions`)

<img width="795" alt="Screen Shot 2021-12-08 at 1 23 11 PM" src="https://user-images.githubusercontent.com/13460330/145286533-c341eb46-0779-4d41-b122-f9bb3e21a05e.png">

<table>
<tr>
	<td>Status
	<td>Task</td>
	<td>Complexity
	<td>Uncertainty
	<td>Expected
	<td>Worst-case
	<td>Notes
<tr>
	<td>
	<td>Move properties stats into corresponding events
	<td>small
	<td>moderate
	<td>1d
	<td>1.5d
	<td>
<tr>
	<td>
	<td>🎨 Property-level validation rule building
	<td>medium
	<td>moderate
	<td>3d
	<td>4.5d
	<td>Blocked on mockups and backend rules functionality. Allowing users to build rules (i.e., `email` property contains `@posthog.com`) to decide which events should be accepted or not. Probably comes with a reasonable default definition.
<tr>
	<td>
	<td>Link event property to corresponding insight
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td>
<tr>
	<td>
	<td>Add sortable columns to properties table for taxonomic metrics
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td>blocked on backend logic to calculate properties stats
</table>

`Event Log` Tab 🦄 
<table>
<tr>
	<td>Status
	<td>Task</td>
	<td>Complexity
	<td>Uncertainty
	<td>Expected
	<td>Worst-case
	<td>Notes
<tr>
	<td>
	<td>Beef up events log filter
	<td>medium
	<td>moderate
	<td>2d
	<td>3d
	<td> Mockup -  #5546
<tr>
	<td>
	<td>Link event to event definition
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td> Mockup -  #5546
<tr>
	<td>
	<td>Link event property to event property definition
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td> Mockup - #5546
</table>

Insight Trends 🦄
<table>
<tr>
	<td>
	<td>Task</td>
	<td>Complexity
	<td>Uncertainty
	<td>Expected
	<td>Worst-case
	<td>Notes
<tr>
	<td>
	<td>Allow filtering by `event property`
	<td>small
	<td>extreme
	<td>1d
	<td>5d
	<td>
</table>

Taxonomic Filter 🦄

Blocked on finishing `Event Definitions` tab.

<table>
<tr>
	<td>Status
	<td>Task</td>
	<td>Complexity
	<td>Uncertainty
	<td>Expected
	<td>Worst-case
	<td>Notes
<tr>
	<td>
	<td>Expose counting stats for all events and event properties
	<td>small
	<td>moderate
	<td>1d
	<td>1.5d
	<td>blocked on all Event Definitions
<tr>
	<td>
	<td>Link events to event definitions
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td>
<tr>
	<td>
	<td>Only showing properties that are relevant/can be filtered on
	<td>small
	<td>moderate
	<td>1d
	<td>1.5d
	<td>
<tr>
	<td>
	<td>~~Display stale tags for events that were decommissioned~~
	<td>small
	<td>moderate
	<td>1d
	<td>1.5d
	<td>Already live
</table>

Automated event + properties calculations 🐴 

<table>
<tr>
	<td>Status
	<td>Task</td>
	<td>Complexity
	<td>Uncertainty
	<td>Expected
	<td>Worst-case
	<td>Notes
<tr>
	<td>
	<td>Cron to run calculations every hour to compute event + properties
	<td>medium
	<td>moderate
	<td>2d
	<td>3d
	<td>
<tr>
	<td>
	<td>"Bad events" rule checking, counting, and handling
	<td>medium
	<td>moderate
	<td>3d
	<td>4.5d
	<td>
<tr>
	<td>
	<td>Active/Inactive events update API
	<td>small
	<td>low
	<td>1d
	<td>1.1d
	<td>
<tr>
	<td>
	<td>🔻 Backfilling counting stats for events up to last X days
	<td>medium
	<td>high
	<td>3d
	<td>6d
	<td>
</table>

## An argument for removing `Properties Stats` tab altogether and merging with `Events Definitions` (see above).
1. How much value does a separate properties tab bring? 
	* At first glance, the event properties table is a collection of all properties across all events that are ingested. My hypothesis is that a user is coming to the event properties table with the question of "I know what events I'm interested in. Now tell me which events fired for this property and how often." We can solve this by linking back and forth between the events and properties tables but I believe we can also rid of confusion altogether by removing this mental hoop. Better yet, if insight trend taxonomic filters support counting by event properties, we would have a first class way of showing property stats in a more meaningful way.

**Comment by macobo** [comment]
Note: Also worth thinking about group level properties now with tab per each group type. :)

**Comment by paolodamico** [comment]
This is a bit hard to follow. My suggestion would be to create an issue for each item (the top priorities) and then reference them in #7581, ordering them by priorities. That way everyone on the team can provide feedback on a) prioritization and b) the specific solution details. Further, it will be clear where we'll need design help.

Some tactical feedback on the items outlined (but let's continue the conversation on each issue).
- For me the absolute biggest thing we can do is figure out the right UX to display taxonomy information when creating insights / actions. And we'll definitely need design input.
- Sounds like one of the top priorities is the data type recognition. Worth syncing up with @mariusandra re his work on this.
- Stale events is mostly done.
- What are we trying to solve with event version history?
- Sounds like a reasonable assumption re property definitions, let's put a full design of this in front of users and get some feedback. We could also check the proportion of properties vs. event definitions used currently.

**Comment by alexkim205** [comment]
Closing this out in favor of #8203
